src/x/opentelemetry/config.go

```diff
@@ -0,0 +1,136 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package opentelemetry provides Open Telemetry configuration.
+package opentelemetry
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/uber-go/tally"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpgrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+	"google.golang.org/grpc"
+)
+
+// Configuration configures an OpenTelemetry trace provider.
+type Configuration struct {
+	ServiceName string            `yaml:"serviceName"`
+	Endpoint    string            `yaml:"endpoint"`
+	Insecure    bool              `yaml:"insecure"`
+	Attributes  map[string]string `yaml:"attributes"`
+}
+
+// TracerProviderOptions is a set of options to use when creating the
+// trace provider.
+type TracerProviderOptions struct {
+	// Attributes is a set of programmatic attributes to add at construction.
+	Attributes []attribute.KeyValue
+}
+
+// NewTracerProvider returns a new tracer provider.
+func (c Configuration) NewTracerProvider(
+	ctx context.Context,
+	scope tally.Scope,
+	opts TracerProviderOptions,
+) (*sdktrace.TracerProvider, error) {
+	attributes := make([]attribute.KeyValue, 0, 1+len(c.Attributes)+len(opts.Attributes))
+	attributes = append(attributes, semconv.ServiceNameKey.String(c.ServiceName))
+	for k, v := range c.Attributes {
+		attributes = append(attributes, attribute.String(k, v))
+	}
+	attributes = append(attributes, opts.Attributes...)
+
+	res, err := resource.New(ctx, resource.WithAttributes(attributes...))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	driverOpts := []otlpgrpc.Option{
+		otlpgrpc.WithEndpoint(c.Endpoint),
+		otlpgrpc.WithDialOption(grpc.WithBlock()),
+	}
+	if c.Insecure {
+		driverOpts = append(driverOpts, otlpgrpc.WithInsecure())
+	}
+	driver := otlpgrpc.NewDriver(driverOpts...)
+	traceExporter, err := otlp.NewExporter(ctx, driver)
+	if err != nil {
+		return nil, fmt.Errorf("failed to trace exporter: %w", err)
+	}
+
+	// Register the trace exporter with a TracerProvider, using a batch
+	// span processor to aggregate spans before export.
+	batchSpanProcessor := sdktrace.NewBatchSpanProcessor(traceExporter)
+	tracerMetricsProcessor := newTraceSpanProcessor(scope)
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(batchSpanProcessor),
+		sdktrace.WithSpanProcessor(tracerMetricsProcessor),
```

Comment:
Is this intended?

Reply:
Yeah, basically this adds metrics about the spans being created/completed.